File: ios/HybridSquircleNativeView.swift
import NitroModules
import UIKit

final class HybridSquircleNativeView: HybridSquircleNativeViewSpec, RecyclableView {
  let view = SquircleRenderView()

  private let stateLock = NSLock()
  private var state = SquircleRenderState()
  private var batching = false

  var cornerSmoothing: Double {
    get { read(\.cornerSmoothing) }
    set { write(\.cornerSmoothing, newValue) }
  }
  var squircleBackgroundColor: Double {
    get { read(\.backgroundColor) }
    set { write(\.backgroundColor, newValue) }
  }
  var squircleBorderColor: Double {
    get { read(\.borderColor) }
    set { write(\.borderColor, newValue) }
  }
  var squircleBorderWidth: Double {
    get { read(\.borderWidth) }
    set { write(\.borderWidth, newValue) }
  }
  var squircleBorderStyle: SquircleBorderStyle {
    get { read(\.borderStyle) }
    set { write(\.borderStyle, newValue) }
  }
  var topLeftRadius: Double {
    get { read(\.topLeftRadius) }
    set { write(\.topLeftRadius, newValue) }
  }
  var topRightRadius: Double {
    get { read(\.topRightRadius) }
    set { write(\.topRightRadius, newValue) }
  }
  var bottomRightRadius: Double {
    get { read(\.bottomRightRadius) }
    set { write(\.bottomRightRadius, newValue) }
  }
  var bottomLeftRadius: Double {
    get { read(\.bottomLeftRadius) }
    set { write(\.bottomLeftRadius, newValue) }
  }
  var overflowHidden: Bool {
    get { read(\.overflowHidden) }
    set { write(\.overflowHidden, newValue) }
  }
  var squircleShadowColor: Double {
    get { read(\.shadowColor) }
    set { write(\.shadowColor, newValue) }
  }
  var squircleShadowOpacity: Double {
    get { read(\.shadowOpacity) }
    set { write(\.shadowOpacity, newValue) }
  }
  var squircleShadowRadius: Double {
    get { read(\.shadowRadius) }
    set { write(\.shadowRadius, newValue) }
  }
  var shadowOffsetX: Double {
    get { read(\.shadowOffsetX) }
    set { write(\.shadowOffsetX, newValue) }
  }
  var shadowOffsetY: Double {
    get { read(\.shadowOffsetY) }
    set { write(\.shadowOffsetY, newValue) }
  }

  func beforeUpdate() {
    stateLock.lock()
    batching = true
    stateLock.unlock()
  }

  func afterUpdate() {
    stateLock.lock()
    batching = false
    let snapshot = state
    stateLock.unlock()
    apply(snapshot)
  }

  func prepareForRecycle() {
    stateLock.lock()
    state = SquircleRenderState()
    batching = false
    stateLock.unlock()
    if Thread.isMainThread {
      view.reset()
    } else {
      DispatchQueue.main.async { [weak view] in view?.reset() }
    }
  }

  private func read<Value>(_ keyPath: KeyPath<SquircleRenderState, Value>) -> Value {
    stateLock.lock()
    defer { stateLock.unlock() }
    return state[keyPath: keyPath]
  }

  private func write<Value>(_ keyPath: WritableKeyPath<SquircleRenderState, Value>, _ value: Value) {
    stateLock.lock()
    state[keyPath: keyPath] = value
    let shouldApply = !batching
    let snapshot = state
    stateLock.unlock()
    if shouldApply {
      apply(snapshot)
    }
  }

  private func apply(_ snapshot: SquircleRenderState) {
    if Thread.isMainThread {
      view.apply(snapshot)
    } else {
      DispatchQueue.main.async { [weak view] in view?.apply(snapshot) }
    }
  }
}

File: android/src/main/java/com/margelo/nitro/nitrosquircle/HybridSquircleNativeView.kt
package com.margelo.nitro.nitrosquircle

import android.os.Looper
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.views.RecyclableView

internal class HybridSquircleNativeView(context: ThemedReactContext) :
  HybridSquircleNativeViewSpec(), RecyclableView {
  override val view = SquircleRenderView(context)

  private val state = SquircleRenderState()
  private var batching = false

  override var cornerSmoothing: Double
    get() = synchronized(state) { state.cornerSmoothing }
    set(value) = update { cornerSmoothing = value }
  override var squircleBackgroundColor: Double
    get() = synchronized(state) { state.backgroundColor }
    set(value) = update { backgroundColor = value }
  override var squircleBorderColor: Double
    get() = synchronized(state) { state.borderColor }
    set(value) = update { borderColor = value }
  override var squircleBorderWidth: Double
    get() = synchronized(state) { state.borderWidth }
    set(value) = update { borderWidth = value }
  override var squircleBorderStyle: SquircleBorderStyle
    get() = synchronized(state) { state.borderStyle }
    set(value) = update { borderStyle = value }
  override var topLeftRadius: Double
    get() = synchronized(state) { state.topLeftRadius }
    set(value) = update { topLeftRadius = value }
  override var topRightRadius: Double
    get() = synchronized(state) { state.topRightRadius }
    set(value) = update { topRightRadius = value }
  override var bottomRightRadius: Double
    get() = synchronized(state) { state.bottomRightRadius }
    set(value) = update { bottomRightRadius = value }
  override var bottomLeftRadius: Double
    get() = synchronized(state) { state.bottomLeftRadius }
    set(value) = update { bottomLeftRadius = value }
  override var overflowHidden: Boolean
    get() = synchronized(state) { state.overflowHidden }
    set(value) = update { overflowHidden = value }
  override var squircleShadowColor: Double
    get() = synchronized(state) { state.shadowColor }
    set(value) = update { shadowColor = value }
  override var squircleShadowOpacity: Double
    get() = synchronized(state) { state.shadowOpacity }
    set(value) = update { shadowOpacity = value }
  override var squircleShadowRadius: Double
    get() = synchronized(state) { state.shadowRadius }
    set(value) = update { shadowRadius = value }
  override var shadowOffsetX: Double
    get() = synchronized(state) { state.shadowOffsetX }
    set(value) = update { shadowOffsetX = value }
  override var shadowOffsetY: Double
    get() = synchronized(state) { state.shadowOffsetY }
    set(value) = update { shadowOffsetY = value }

  override fun beforeUpdate() {
    synchronized(state) { batching = true }
  }

  override fun afterUpdate() {
    val snapshot = synchronized(state) {
      batching = false
      state.copy()
    }
    apply(snapshot)
  }

  override fun prepareForRecycle() {
    synchronized(state) {
      state.copyFrom(SquircleRenderState())
      batching = false
    }
    view.post { view.reset() }
  }

  override fun onDropView() {
    view.post { view.close() }
  }

  private inline fun update(change: SquircleRenderState.() -> Unit) {
    val snapshot = synchronized(state) {
      state.change()
      if (batching) null else state.copy()
    }
    if (snapshot != null) apply(snapshot)
  }

  private fun apply(snapshot: SquircleRenderState) {
    if (Looper.myLooper() == Looper.getMainLooper()) {
      view.apply(snapshot)
    } else {
      view.post { view.apply(snapshot) }
    }
  }

  private fun SquircleRenderState.copyFrom(other: SquircleRenderState) {
    cornerSmoothing = other.cornerSmoothing
    backgroundColor = other.backgroundColor
    borderColor = other.borderColor
    borderWidth = other.borderWidth
    borderStyle = other.borderStyle
    topLeftRadius = other.topLeftRadius
    topRightRadius = other.topRightRadius
    bottomRightRadius = other.bottomRightRadius
    bottomLeftRadius = other.bottomLeftRadius
    overflowHidden = other.overflowHidden
    shadowColor = other.shadowColor
    shadowOpacity = other.shadowOpacity
    shadowRadius = other.shadowRadius
    shadowOffsetX = other.shadowOffsetX
    shadowOffsetY = other.shadowOffsetY
  }
}
